client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,417 @@
+}
+
+/* Get encrypted session key (enctypted with remote key) */
+kaa_error_t ext_get_encrypted_session_key(uint8_t **buffer, size_t *buffer_size,
+        const uint8_t *remote_key, size_t remote_key_size)
+{
+    if (!buffer || !buffer_size || !remote_key || !remote_key_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    mbedtls_pk_context local_ctx;
+    mbedtls_pk_init(&local_ctx);
+    int err = set_rsa_public_key(&local_ctx, remote_key, remote_key_size);
+    if (err) {
+        goto exit;
+    }
+    err = rsa_encrypt(&local_ctx, keys.session_key,
+            KAA_SESSION_KEY_LENGTH, keys.encrypted_session_key);
+    if (err) {
+        goto exit;
+    }
+
+    keys.encrypted_session_key_length = ENC_SESSION_KEY_LENGTH;
+
+    *buffer = keys.encrypted_session_key;
+    *buffer_size = keys.encrypted_session_key_length;
+
+exit:
+    mbedtls_pk_free(&local_ctx);
+    return err ? KAA_ERR_BADDATA : KAA_ERR_NONE;
+}
+
+size_t ext_get_encrypted_data_size(size_t input_size)
+{
+    if (!input_size) {
+        return 0;
+    }
+
+    if (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE != 0) {
+        return AES_ECB_ENCRYPTION_CHUNCK_SIZE -
+               (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE) + input_size;
+    }
+
+    return input_size + AES_ECB_ENCRYPTION_CHUNCK_SIZE;
+}
+
+static int rsa_sign(mbedtls_pk_context *pk, const uint8_t *input,
+        size_t input_size, uint8_t *output, size_t *output_size)
+{
+    int ret = 0;
+    uint8_t hash[32];
+    mbedtls_entropy_context entropy;
+    mbedtls_ctr_drbg_context ctr_drbg;
+    const uint8_t pers[] = "mbedtls_pk_sign";
+
+    mbedtls_entropy_init(&entropy);
+    mbedtls_ctr_drbg_init(&ctr_drbg);
+
+
+    if ((ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy,
+            pers,sizeof(pers) - 1)) != 0) {
+        goto exit;
+    }
+
+    const mbedtls_md_info_t *info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
+
+    if ((ret = mbedtls_md(info, input, input_size, hash)) != 0) {
+        goto exit;
+    }
+
+    ret = mbedtls_pk_sign(pk, MBEDTLS_MD_SHA1, hash, 0, output, output_size,
+            mbedtls_ctr_drbg_random, &ctr_drbg);
+
+exit:
+    mbedtls_ctr_drbg_free(&ctr_drbg);
+    mbedtls_entropy_free(&entropy);
+
+    return ret;
+}
+
+
+kaa_error_t ext_get_signature(const uint8_t *input, size_t input_size,
+                              uint8_t **output, size_t *output_size)
+{
+    if (!input || !input_size || !output || !output_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    kaa_error_t error = rsa_sign(&pk_context_, input, input_size, keys.signature, output_size);
+
+    if (error) {
+        return KAA_ERR_BADDATA;
+    }
+    *output = keys.signature;
+    *output_size = KAA_SIGNATURE_LENGTH;
+
+    return KAA_ERR_NONE;
+}
+
+void ext_get_sha1_public(uint8_t **sha1, size_t *length)
+{
+    if (!sha1 || !length) {
+        return;
+    }
+    static uint8_t sha1_public[KAA_SHA1_PUB_LEN];
+    static int initialization = false;
```

Comment:
Logically, sentence `SHA key is not initialization` is wrong. Correct is `SHA key is not initialized.` Use word `initialized` or `initialization_done` or something like this.